proposals/generic-ref-struct.md

```diff
@@ -0,0 +1,424 @@
+Unconstraining our generics
+=====
+
+## Summary
+This proposal is an aggregation of several different proposals for making `ref struct` more generally usable in our type system. The goal is to unify code paths today that have to specialize on `ref struct` and other types due to the inability of `ref struct` to participate in generics and implement interfaces. This proposal will allow us to generalize on `ref struct` as naturally as other types. 
+
+## Motivation
+This feature will allow developers to remove `ref struct` specific code by instead letting them implement interfaces that are used in generic contexts to generalize algorithms. For example an explicit goal is to allow `ref struct` to implement `ISpanFormattable`. The `ISpanFormattable` interface and `ref struct` are used in low level perf sensitive areas but our current language limitations prevent them from being used together.
+
+This will also allow for combinations like `Span<Span<T>>`. This has long been a desire of the runtime team, who often want to represent collections of `Span<T>` but are unable to do so due to our current limitations.
+
+It also will enable the creation of generic collections of `ref struct`. For instance having a `List<T>` style [data structure](#ref-list) with a hard capacity limit will be possible with the features in this proposal.
+
+## Detailed Design 
+
+### ref struct generic arguments
+The compiler will allow type parameters to be (un)constrained such that they can accept `ref struct` as arguments using the `where T : ref struct` syntax. 
+
+```c#
+void Swap<T>(ref T left, ref T right)
+    where T : ref struct
+{
+    T temp = left;
+    left = right;
+    right = temp;
+}
+
+Span<byte> span1 = new byte[4];
+Span<byte> span2 = default;
+
+// Can use ref struct type arguments 
+Swap<Span<byte>>(ref span1, ref span2);
+
+// Can also use normal type arguments
+Swap<byte>(ref span1[0], ref span1[1]);
+```
+
+At the call site this relaxes the set of legal type arguments to included `ref struct` *in addition* to the types already allowed. Even though this appears in the constraint position it serves more of an anti-constraint because it expands the set of possible types. 
+
+As an expansion it means that `where T : ref struct` is not compatible with type parameters lacking the same annotation:
+
+```c#
+void M1<U>() { }
+void M2<T>()
+    where T : ref struct
+{
+    M1<T>(); // Error: type parameter T is not compatible with U because it does not 
+             // allow `ref struct`
+}
+```
+
+At the implementation all type parameters having the `where T : ref struct` annotation are treated as if the type is a `ref struct`. This means the type parameter has all of the limitations that would be expected a `ref struct` type. For example:
+
+- Instances of `T` cannot be boxed or appear on the heap
+- Instances of `T` can only appear as fields inside a `ref struct`
```

Comment:
We could loosen this restriction and make ordinary generic structs be seen as `ref struct`s if at least one of their generic parameters (and perhaps but not necessarily) that is stored as a field is either a `ref struct`, or a generic parameter that can be a `ref struct`.

For an example, we could redefine `ValueTuple<T1, T2>` like the following:

``` csharp
public struct ValueTuple<T1, T2> where T1 : ref struct, T2 : ref struct
{
    public T1 Item1;
    public T2 Item2;
}
```

and `(int, ReadOnlySpan<int>)` would be allowed and be restricted by the compiler to live only on the stack. All other value tuples would still be allowed to reside on the heap, including generic ones, unless their generic parameters are allowed to be `ref struct`s.

Reply:
Some version of this might be possible but I was concerned about how such a struct could implement interfaces, for example. Would we need to go through a bunch of built-in generic types and add `where T : ref struct` to make it all work?

I haven't thought very deeply about this so I anticipate there could be more stumbling blocks here that make it more difficult than just "let ValueTuple contain ref structs".

Reply:
Oh, that's right, we couldn't implement `ITuple` on value tuples for example.